text/0000-additional-exported-layers.md

```diff
@@ -0,0 +1,86 @@
+# Meta
+[meta]: #meta
+- Name: Additional exportable layers
+- Start Date: 2021-08-13
+- Author(s): [@samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Allow users to provide additional volumes that can be exported similar to the application directory and allow easy configuration of the application directory. These volumes are defined via the `buildpack.exportable-volumes` by each buildpack.
+
+# Definitions
+[definitions]: #definitions
+
+- `exportable-volumes`: A set of volumes describing where the buildpack process is likely to write data specific and should be exported alongside the application workspace.
+
+# Motivation
+[motivation]: #motivation
+
+The main motivation behind this RFC is to unlock exporting application images that require libraries or software to be installed in certain directories apart from the default application directory and these installations are rebase safe and self-contained. This unlocks use cases where buildpacks may install software in `/opt` directories (for AWS Lambda extensions or other common standalone software) for example or preserve some settings in a `/home/$USER/.config` for the output application. This provides a rebase safe extension point without major changes to the buildpacks API.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Buildpack authors would be able to buildpacks with `buildpack.exportable-volumes` keys in their `buildpack.toml` files.
+
+For example a a buildpack may look like -
+
+```toml
+api = "<buildpack API version>"
+
+[buildpack]
+id = "<buildpack ID>"
+name = "<buildpack name>"
+exportable-volumes = ["aws-extensions", "user-config-dir"]
+```
+
+The above would only be valid for normal buildpacks and not meta-buildpacks. Based on the `exportable-volumes` field, a platform would be responsible for validating that the `build` and `run` image do not have any content on these paths. By default these volume names will map to `/workspaces/<exportable-volume-name>` but a user may specify a label `io.buildpacks.exportable-volumes` on the `build` image to map a volume name to a different location. This label would look like -
+
+```json
+[
+    {"name": "user-config-dir", "value": "/home/cnb/.config"},
```

Comment:
I like this mapping concept being done by the base image author. 

Could this list be something buildpack authors could get at and then name layers that match `user-config-dir` that gets exported to `/home/cnb/.config` during export phase? Or perhaps a key in the `layer.toml` has to be set so it is more explicit.

The requirement for an empty dir and symlinking for builder could potentially be avoided. A buildpack author could put a new binary in the resulting app image's `/usr/bin` by putting the binary in `/layers/<bp-dir>/usr-bin/`. If a new stack is used that doesn't support it, the layer would still exist on the resulting app image but not in the overridden path.

Giving the buildpack a dictionary so they can easily check if "usr/bin" is available to use would allow buildpack authors to either put the binary in "usr/bin" or do more work like altering PATH and whatnot to get it to work on other base images.

```
{
  "usr/bin" : "usr-bin",
  "~/.config": "user-config-dir"
}
```